tools/backups2datalad/zarr.py

```diff
@@ -0,0 +1,405 @@
+                self.rmtree(p)
+            else:
+                log.info("Zarr %s: deleting %s", self.zarr_id, p)
+                p.unlink()
+                self.report.deleted += 1
+        dirpath.rmdir()
+
+    def prune_deleted(self) -> None:
+        log.info("Zarr %s: deleting extra files", self.zarr_id)
+        dirs = deque([self.repo])
+        empty_dirs: deque[Path] = deque()
+        while dirs:
+            d = dirs.popleft()
+            is_empty = True
+            for p in list(d.iterdir()):
+                path = p.relative_to(self.repo).as_posix()
+                if d == self.repo and p.name in (
+                    CHECKSUM_FILE,
+                    ".datalad",
+                    ".git",
+                    ".gitattributes",
+                ):
+                    pass
+                elif p.is_dir():
+                    dirs.append(p)
+                elif path not in self.extant_paths:
+                    log.info("Zarr %s: deleting %s", self.zarr_id, path)
+                    p.unlink()
+                    self.report.deleted += 1
+                else:
+                    is_empty = False
+            if is_empty and d != self.repo:
+                empty_dirs.append(d)
+        while empty_dirs:
+            d = empty_dirs.popleft()
+            d.rmdir()
+            if d.parent != self.repo and not any(d.parent.iterdir()):
+                empty_dirs.append(d.parent)
+
+    async def aiter_file_entries(self) -> AsyncIterator[RemoteZarrEntry]:
+        root = RemoteZarrEntry(client=None, zarr_id=self.zarr_id, parts=())
+        dirs = deque([root])
+        async with httpx.AsyncClient() as client:
+            while dirs:
+                d = dirs.popleft()
+                path = "".join(p + "/" for p in d.parts)
+                log.debug(
+                    "Zarr %s: Getting listing for %s",
+                    self.zarr_id,
+                    path or "root directory",
+                )
+                try:
+                    r = await arequest(
+                        client, "GET", f"{self.api_url}/zarr/{self.zarr_id}.zarr/{path}"
+                    )
+                except httpx.HTTPStatusError as e:
+                    if e.response.status_code == 404 and d.is_root():
+                        # Empty Zarr
+                        return
+                    else:
+                        raise
+                listing = ZarrListing.parse_obj(r.json())
+                for name in listing.dirnames:
+                    dirs.append(
+                        d._get_subpath(
+                            name, isdir=True, checksum=listing.checksums[name]
+                        )
+                    )
+                for name in listing.filenames:
+                    yield d._get_subpath(
+                        name, isdir=False, checksum=listing.checksums[name]
+                    )
+
+    async def stat(self, entry: RemoteZarrEntry) -> Tuple[str, ZarrEntryStat]:
```

Comment:
oh -- we do it per each key? that might be expensive/slow, isn't it?
is there some API to query a number of keys at once  or it would just be serialized into individual ones? (hard to believe, didn't check S3 API)

Reply:
There does not appear to be an API for querying batches of keys from S3.  I'm not clear on what you mean by "it would just be serialized into individual ones".

Reply:
Unless it could/would be implemented quickly - I think it is ok to proceed here *as is* for now, but in the long run we will need more efficient API for that.
Please file an issue (or PR if you see a straightforward way to implement) with dandi-archive to expose necessary information in dandi-api without requiring per file in Zarr querying -- such querying would be of huge (and avoidable) impact on dandi-api!  Another part of the motivation for that is that `/zarr/{zarr_id}.zarr/{path}` is there to provide a "sparse" access to zarr for zarr clients, so they would unlikely be traversing the entire zarr folder.  In our case, we would.

Possibly related is https://github.com/dandi/dandi-archive/issues/925 -- as a "mid point" solution could be that information is exposed in that `zarr_content_read` endpoint, but since it is not recursive, most likely it would be a good fit.

Reply:
> Moreover, note that just getting the S3 key for a Zarr entry would still require an individual `HEAD` request to see where the `/zarr/{zarr_id}.zarr/{path}` endpoint redirects to.

worse comes to worse (no information on redirect to S3 URL) we can hardcode the assumption of 1-to-1 mapping into S3 bucket.  I believe zarr implementation doesn't care about versionId's ATM, does it? So we would need to get a versionId'ed via S3 request anyways :-/

Reply:
I posted a comment on that issue.

Reply:
With all that in mind -- should we may be just not go through API for this at all and just go straight to the S3 bucket and get "current version" (if multiple versions present for some keys) from S3 and be done?
Because otherwise we can run into a problem of dandi-api not knowing current exact situation in S3 (zarr is being uploaded), us needing to crash when encountering some incongruency (e.g. key was deleted) or trying to figure out VersionId if multiple versions are present and the most recent one not corresponding in size to what dandi-api knows?

We can still register URLs pointing to our API similarly to how we do for assets.

Reply:
Could you elaborate on how you're envisioning this new implementation?

Reply:
> @yarikoptic No, I did not test on any Zarrs in the main archive. It's my understanding that all of the Zarrs on production are huge.

yes and they would be our target. unfortunately as I have discovered (and as I asked in slack) their metadata is not updated yet, so they all "size 0" while @satra uploading using s5cmd I believe.  Did you try on some sizeable on staging?

Reply:
The only tests I ran were for the actual test code, which creates some small Zarrs in staging.

Reply:
please run then on some sizeable zarrs in some dandiset in staging.  May be with all per-file querying it would should to be infeasible or bring down the staging server to its knees.

Reply:
Wouldn't this rely on the assumption that an entry `foo/bar/baz` in a Zarr is located at `/zarr/{zarr_id}/foo/bar/baz` on S3?  That strikes me as an implementation detail.

Reply:
if we are considering other tools, why not just use `s5cmd` instead of datalad to crawl the prefix (it's significantly faster than for that kind of ls, even with etag retrieval.

Reply:
we need to register keys and URLs for those files in annex... actually ideally not even download the files in that step (forgot if we paid attention to that here), but just mint/register based on ETag in git-annex similarly to how we do it for regular assets, and then download/backup in a separate step.  I don't think that "crawl"ing is the one which would be bottleneck.  It would be the rest of interaction with git-annex + making update efficient (i.e. instrument interaction with S3 to only react on new stuff as datalad-crawler does, although I believe it does get a list of all keys/versions for that first... would need to check).
So -- not sure if use of s5cmd would mitigate anything here.